### PR TITLE
Get rid of __init__(None)

### LINKED ISSUE #38: Get rid of the __init__(None) idiom
Currently, there are lots of code looking like these
```python
cdef class Device:
    """...
    Parameters
    ----------
    name : Optional[str], ...
        The name of the playback device.  If it is `None`,
        the object is left uninitialized.
    ...
    """
    cdef alure.Device impl
    def __init__(self, name: Optional[str] = '', ...) -> None:
        if name is None: return
        ...

cdef Device device = Device(None)
device.impl = ...
```
which is (needless to say) stupid (thank to me who invented it, ah, I'm so proud of myself) and error-prone: imagine users somehow call it this way and get segfault all over the place.  I now we're all adults and stuff, but, ehm, how to say, if only Python does not offer a solution like so
```python
cdef class Device:
    """...
    Parameters
    ----------
    name : str, ...
        The name of the playback device.
    ...
    """
    cdef alure.Device impl
    def __init__(self, name: str = '', ...) -> None:
        ...

cdef Device device = Device.__new__(Device)
device.impl = ...
```

@9a24f0, since you haven't touched the codebase for so long, maybe warming up by solving this might be a good idea.

### PR DESCRIPTION
**Pull request by 9a24f0**

### CODE REVIEW

File: src/palace.pyx [hunk 1]
```diff
@@ -179,9 +178,8 @@ cdef class Device:
     """
     cdef alure.Device impl
 
-    def __init__(self, name: Optional[str] = '',
+    def __init__(self, name: str = '',
                  fail_safe: bool = False) -> None:
```
Comment (McSinyx) [inline src/palace.pyx]: These 2 lines are short enough to be joined into one.

**Review by McSinyx — CHANGES_REQUESTED** [review]
There are a lot more instances of init call with `None` as the sole argument.  Travis will tell you more.

Edit: I was wrong, Travis doesn't tell enough other than segfault tests.  Keep searching for `(None)` until per's happy.

File: src/palace.pyx (hunk 1 above)
Comment (9a24f0) [reply src/palace.pyx]: this is 72 as I calculated, is it OK for generating docs?

Comment (McSinyx) [reply src/palace.pyx]: 72 is the limit for (multiline) docstrings, for codes limit to 79.

Comment (9a24f0) [reply src/palace.pyx]: Oof, it is 73, but I merged it either

Comment (McSinyx) [reply src/palace.pyx]: It's code so it's OK.  BTW read the comment above and keep trying till the tests pass (run `tox` locally to save time).

Comment (9a24f0) [reply src/palace.pyx]: But how could I fix those, simply delete that?

Comment (McSinyx) [reply src/palace.pyx]: `Source(None)` should be changed into `Source.__new__(Source)` as you did for some of them.

File: src/palace.pyx [hunk 2]
```diff
@@ -689,14 +686,12 @@ cdef class Source:
...
+    context : Context
         The context from which the source is to be created.
-        If it is `None`, the object is left uninitialized.
     """
     cdef alure.Source impl
 
-    def __init__(self, context: Optional[Context]) -> None:
-        if context is None: return
+    def __init__(self, context: Context) -> None:
         self.impl = (<Context> context).impl.create_source()
```
Comment (McSinyx) [inline src/palace.pyx]: Notice the unnecessary casting here.

File: src/palace.pyx [hunk 3]
```diff
@@ -1237,14 +1232,12 @@ cdef class SourceGroup:
...
+    context : Context
         The context from which the source group is to be created.
-        If it is `None`, the object is left uninitialized.
     """
     cdef alure.SourceGroup impl
 
-    def __init__(self, context: Optional[Context]) -> None:
-        if context is None: return
+    def __init__(self, context: Context) -> None:
         self.impl = (<Context> context).impl.create_source_group()
```
Comment (McSinyx) [inline src/palace.pyx]: Here too.

**Review by McSinyx — CHANGES_REQUESTED** [review]

File: src/palace.pyx (hunk 2 above)
Comment (9a24f0) [reply src/palace.pyx]: so we dont cast it to None?

Comment (McSinyx) [reply src/palace.pyx]: What, no, I mean `context` needn't be casted (`(<Context> context)`).

File: src/palace.pyx (hunk 1 above)
Comment (9a24f0) [reply src/palace.pyx]: I did the changes but when I compile, it fails

**Comment by McSinyx** [comment]
You have to eliminate **all** uses of `ClassName(None)` in order for the tests to pass.

**Comment by 9a24f0** [comment]
I search for `(None)` and replaced things, how could I miss any `ClassName(None)`

**Comment by 9a24f0** [comment]
Since this is kinda late, I will commit my changes now and sleep

File: src/palace.pyx (hunk 2 above)
Comment (McSinyx) [reply src/palace.pyx]: Drop the parentheses please.

**Comment by McSinyx** [comment]
> I search for `(None)` and replaced things, how could I miss any `ClassName(None)`

Something is definitely wrong with your editor:  https://github.com/McSinyx/palace/blob/961b508106665a6031176ded4a5498bbbf330b06/src/palace.pyx

**Comment by 9a24f0** [comment]
There are 5 of them now, but when I change it from `ClassName(None)` to `ClassName.__new__(Classname)` it fails the test

**Comment by McSinyx** [comment]
Push the changes up so I can view them please.

**Comment by 9a24f0** [comment]
there are 5 of them in my prev push, I've checked and when I change any of them the test fails

**Comment by McSinyx** [comment]
But the test still fails without changing them.  Could you please push the changes so I can view them?

**Comment by McSinyx** [comment]
I see, the test fails because Cython is not able to infer the type of the `impl` and think it's just accessing a normal Python attribute, while in fact we're setting an extension one.  This can be fixed by declaring the type of the variable (at the beginning of the method), e.g.
```diff
diff --git a/src/palace.pyx b/src/palace.pyx
index e7f0d0b..e4387c0 100644
--- a/src/palace.pyx
+++ b/src/palace.pyx
@@ -649,6 +649,7 @@ cdef class Buffer:
     @property
     def sources(self) -> List[Source]:
         """`Source` objects currently playing the buffer."""
+        cdef Source source
         sources = []
         for alure_source in self.impl.get_sources():
             source = Source.__new__(Source)
@@ -781,7 +782,7 @@ cdef class Source:
         --------
         SourceGroup : A group of `Source` references
         """
-        source_group = SourceGroup.__new__(SourceGroup)
+        cdef SourceGroup source_group = SourceGroup.__new__(SourceGroup)
         source_group.impl = self.impl.get_group()
         return source_group or None
```

File: src/palace.pyx (hunk 2 above)
Comment (McSinyx) [reply src/palace.pyx]: Gentle ping (both cases), otherwise LGTM.

**Review by McSinyx — APPROVED** [review]
LGTM now.  Feel free to **Squash and merge** when you feel too, the button is in the drop down menu down there.

**Pull request merged**